presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -311,47 +406,160 @@ private void processPage(Page page)
-            memorySize -= operatorContext.getOperatorPreAllocatedMemory().toBytes();
-            if (memorySize < 0) {
-                memorySize = 0;
+            if (isBusy()) {
+                return true;
             }
+
+            long memorySize = getSizeInMemory();
             if (partial) {
                 return !operatorContext.trySetMemoryReservation(memorySize);
             }
             else {
-                operatorContext.setMemoryReservation(memorySize);
+                if (isSpillingEnabled()) {
+                    systemMemoryContext.setBytes(memorySize);
+
+                    if (shouldSpill(memorySize)) {
+                        log.info(
+                                "Started spilling memory %s",
+                                DataSize.succinctBytes(memorySize));
+
+                        spillToDisk();
+                    }
+                }
+                else {
+                    operatorContext.setMemoryReservation(memorySize);
+                }
                 return false;
             }
         }
 
-        public Iterator<Page> build()
+        private boolean shouldSpill(long memorySize)
+        {
+            return (maxEntriesBeforeSpill > 0 && groupByHash.getGroupCount() > maxEntriesBeforeSpill)
+                    || (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+        }
+
+        private boolean isSpillingEnabled()
+        {
+            return maxEntriesBeforeSpill > 0 || memorySizeBeforeSpill > 0;
+        }
+
+        public Iterator<Page> buildResult()
+        {
+            checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+            if (!spiller.isPresent()) {
+                return buildResultFromMemory();
+            }
+
+            try {
+                // TODO: don't spill here to disk, instead merge disk content with memory content
+                spillToDisk().get();
+                return mergeFromDisk();
+            }
+            catch (InterruptedException | ExecutionException e) {
+                Thread.currentThread().interrupt();
+                throw Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            closeSpiller();
+        }
+
+        private CompletableFuture<?> spillToDisk()
         {
-            List<Type> types = new ArrayList<>(groupByHash.getTypes());
             for (Aggregator aggregator : aggregators) {
-                types.add(aggregator.getType());
+                aggregator.setOutputPartial();
             }
 
+            if (!spiller.isPresent()) {
+                spiller = Optional.of(spillerFactory.get().create(buildTypes()));
+            }
+
+            // we are starting spilling process with current content of the aggregators and groupByHash...
+            busy = spiller.get().spill(buildResultFromMemory(sortedGroupIds(), buildTypes(), groupByHash, aggregators));
```

Comment:
Not quite. This is called during spilling process, so `isSpillInProgress` will return true. This will block `isFull` method (aka `checkFullAndUpdateMemory` method), from updating memory usage until spilling has finished. So memory usage of `groupByHash` and `aggregators` is being tracked (calculated in `getSizeInMemory`).

Since `buildResultFromMemory` returns `Iterator<Page>` and result pages are being generated online (on the fly from `groupyByHash` and `aggregators`), only one single (currently produced) result page is not accounted - but that was true even before my changes and is probably true for every other operator.